streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -923,12 +1090,39 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
                                                    final ValueJoiner<? super V1, ? super V2, ? extends R> joiner,
                                                    final JoinWindows windows,
                                                    final Joined<K1, V1, V2> joined) {
-            final String thisWindowStreamName = builder.newProcessorName(WINDOWED_NAME);
-            final String otherWindowStreamName = builder.newProcessorName(WINDOWED_NAME);
-            final String joinThisName = rightOuter ? builder.newProcessorName(OUTERTHIS_NAME) : builder.newProcessorName(JOINTHIS_NAME);
-            final String joinOtherName = leftOuter ? builder.newProcessorName(OUTEROTHER_NAME) : builder.newProcessorName(JOINOTHER_NAME);
-            final String joinMergeName = builder.newProcessorName(MERGE_NAME);
 
+            final JoinedInternal<K1, V1, V2>  joinedInternal = new JoinedInternal<>(joined);
+            final NamedInternal renamed = new NamedInternal(joinedInternal.name());
+
+            // we actually do not need to generate processor names for join operations if a name is specified.
+            // But, we still need to burn index for all joins operation keep topology backward compatibility.
+            if (renamed.isDefined()) {
+                builder.newProcessorName(WINDOWED_NAME);
+                builder.newProcessorName(WINDOWED_NAME);
+                if (rightOuter)
```

Comment:
nit: add opening and closing curly braces for all the `if` / `else` blocks on lines 1102 to 1108